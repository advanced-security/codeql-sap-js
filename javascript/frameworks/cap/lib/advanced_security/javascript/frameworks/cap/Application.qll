/**
 * Definitions pertaining to the application as a whole.
 */

import javascript
import advanced_security.javascript.frameworks.cap.PackageJson

class RootDirectory extends Folder {
  RootDirectory() {
    exists(PackageJson packageJson | this = packageJson.getJsonFile().getParentContainer())
  }

  /**
   * Gets the path of a file relative to this root directory.
   */
  string getFilePathRelativeToRoot(File file) {
    result = file.getAbsolutePath().regexpReplaceAll(this.getAbsolutePath(), ".") and
    result.charAt(0) = "."
  }

  /**
   * Holds if this root directory of the application contains the given file.
   */
  predicate contains(File file) { exists(this.getFilePathRelativeToRoot(file)) }
}
