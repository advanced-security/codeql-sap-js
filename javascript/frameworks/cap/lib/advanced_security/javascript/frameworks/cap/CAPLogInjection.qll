import javascript
import DataFlow
import advanced_security.javascript.frameworks.cap.CDS

/**
 * A logger obtained by a call to `log` on a CDS facade. Each logger is associated with
 * its unique name.
 */
class CdsLogger extends MethodCallNode {
  string name;

  CdsLogger() {
    this = any(CdsFacade cds).getMember("log").getACall() and
    name = this.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue()
  }
}

/**
 * Arguments of calls to `cds.log.{trace, debug, info, log, warn, error}`
 */
class CdsLogSink extends DataFlow::Node {
  CdsLogSink() {
    exists(CdsLogger log, MethodCallNode loggingMethod |
      this = loggingMethod.getAnArgument() and
      not this.asExpr() instanceof Literal and
      not this.asExpr() instanceof TemplateLiteral and
      loggingMethod.getReceiver().getALocalSource() = log and
      loggingMethod.getMethodName() = ["trace", "debug", "info", "log", "warn", "error"]
    )
  }
}
