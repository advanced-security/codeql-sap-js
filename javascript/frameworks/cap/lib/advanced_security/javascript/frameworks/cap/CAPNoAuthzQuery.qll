import advanced_security.javascript.frameworks.cap.CDS
import advanced_security.javascript.frameworks.cap.CDL
import advanced_security.javascript.frameworks.cap.Conditionals

abstract class CdlElementWithoutAuthn instanceof CdlElement {
  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }
}

abstract class CdlElementWithoutCdsAuthn extends CdlElementWithoutAuthn instanceof CdlElement {
  CdlElementWithoutCdsAuthn() { this.hasNoCdsAccessControl() }
}

class CdlServiceWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlService { }

class CdlEntityWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlEntity {
  CdlEntityWithoutCdsAuthn() {
    this.belongsToServiceWithNoAuthn()
    or
    exists(CdlEntityWithoutCdsAuthn otherCdlEntityWithoutCdsAuthn |
      this.inherits(otherCdlEntityWithoutCdsAuthn)
    )
  }
}

class CdlActionWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlAction {
  CdlActionWithoutCdsAuthn() { this.belongsToServiceWithNoAuthn() }
}

class CdlFunctionWithoutCdsAuthn extends CdlElementWithoutCdsAuthn instanceof CdlAction {
  CdlFunctionWithoutCdsAuthn() { this.belongsToServiceWithNoAuthn() }
}

class CdlElementProtectionWithHandlerRegistration instanceof HandlerRegistration {
  string toString() { result = super.toString() }

  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    super.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
  }

  CdlElementProtectionWithHandlerRegistration() {
    (
      this.isBefore()
      or
      this.isOn()
    ) and
    exists(Handler handler, ConditionalExprOrStatement exprOrStmt |
      handler = this.getHandler() and
      (
        exprOrStmt = handler.getFunction().getBody() or
        exprOrStmt = handler.getFunction().getABodyStmt()
      )
    |
      exprOrStmt.getConditionExpr().getAChildExpr*().flow() instanceof RequestUserIs and
      exists(CdsRequest req |
        exprOrStmt.getPolarity() = true and
        exprOrStmt.getAnElseBranchExpr() = req.getARejectCall().asExpr()
        or
        exprOrStmt.getPolarity() = false and
        exprOrStmt.getAThenBranchExpr() = req.getARejectCall().asExpr()
      )
    )
  }

  string getEntityName() { result = super.getEntityName() }

  string getAnEventName() { result = super.getAnEventName() }
}

abstract class CdlElementWithJsAuthn instanceof CdlElement {
  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }
}

class CdlServiceWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlService {
  CdlServiceWithJsAuthn() {
    exists(CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getAnEventName() = "*"
    )
  }
}

class CdlEntityWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlEntity {
  CdlEntityWithJsAuthn() {
    exists(CdlService service, CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this = service.getAnEntity() and
      service.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getEntityName() = this.getUnqualifiedName()
    )
  }
}

class CdlActionWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlAction {
  CdlActionWithJsAuthn() {
    exists(CdlService service, CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this = service.getAnAction() and
      service.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getAnEventName() = this.getUnqualifiedName()
    )
  }
}

class CdlFunctionWithJsAuthn extends CdlElementWithJsAuthn instanceof CdlFunction {
  CdlFunctionWithJsAuthn() {
    exists(CdlService service, CdlElementProtectionWithHandlerRegistration beforeOrOn |
      this = service.getAFunction() and
      service.getImplementation().getAHandlerRegistration() = beforeOrOn and
      beforeOrOn.getAnEventName() = this.getUnqualifiedName()
    )
  }
}

abstract class CdlElementWithoutJsAuthn extends CdlElementWithoutAuthn instanceof CdlElement { }

class CdlServiceWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlService {
  CdlServiceWithoutJsAuthn() { not this instanceof CdlServiceWithJsAuthn }
}

class CdlEntityWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlEntity {
  CdlEntityWithoutJsAuthn() { not this instanceof CdlEntityWithJsAuthn }
}

class CdlActionWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlAction {
  CdlActionWithoutJsAuthn() { not this instanceof CdlActionWithJsAuthn }
}

class CdlFunctionWithoutJsAuthn extends CdlElementWithoutJsAuthn instanceof CdlFunction {
  CdlFunctionWithoutJsAuthn() { not this instanceof CdlFunctionWithJsAuthn }
}

/**
 * The access to property `user` of a handler's request.
 */
class RequestUser extends SourceNode instanceof PropRef {
  RequestUser() {
    exists(Handler handler |
      this.getBase().getALocalSource() = handler.getRequest() and
      this.getPropertyName() = "user"
    )
  }
}

class RequestUserIs instanceof MethodCallNode {
  string userRole;

  RequestUserIs() {
    exists(RequestUser requestUser |
      this = requestUser.getAMethodCall("is") and
      userRole = this.getArgument(0).getStringValue()
    )
  }

  string toString() { result = super.toString() }
}
