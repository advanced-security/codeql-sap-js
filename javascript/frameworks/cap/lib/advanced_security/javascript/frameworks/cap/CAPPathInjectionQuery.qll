/**
 * Exported functions from CAP `cds.utils`.
 * Functions described from:
 * https://www.npmjs.com/package/@sap/cds?activeTab=code
 */

import javascript
import advanced_security.javascript.frameworks.cap.CDSUtils

abstract class UtilsSink extends DataFlow::Node {
  abstract string sinkType();
}

abstract class UtilsAccessedPathSink extends UtilsSink {
  override string sinkType() { result = "unrestricted file read" }
}

abstract class UtilsControlledDataSink extends UtilsSink {
  override string sinkType() { result = "tainted data being written to a file" }
}

abstract class UtilsControlledPathSink extends UtilsSink {
  override string sinkType() { result = "unrestricted file operations" }
}

/**
 * This represents the data in calls as follows:
 * ```javascript
 * await write ({foo:'bar'}) .to ('some','file.json')
 * ```
 * sinks in this example are:
 * ```javascript
 * {foo:'bar'}
 * ```
 */
class WrittenData extends UtilsControlledDataSink {
  WrittenData() { exists(FileWriters fw | fw.getData() = this) }
}

/**
 * This represents the filepath accessed as an input for the data in calls as follows:
 * ```javascript
 * await copy('db/data').to('dist/db/data')
 * ```
 * sinks in this example are:
 * ```javascript
 * 'db/data'
 * ```
 */
class AccessedPath extends UtilsAccessedPathSink {
  AccessedPath() { exists(FileReaderWriters fw | fw.getFromPath() = this) }
}

/**
 * This represents the filepath where data is written or a file operation is performed in calls as follows:
 * ```javascript
 * await write ({foo:'bar'}) .to ('some','file.json')
 * ```
 * sinks in this example are:
 * ```javascript
 * 'some'
 * 'file.json'
 * ```
 */
class ControlledInputPath extends UtilsControlledPathSink {
  ControlledInputPath() {
    exists(FileReaders fw | fw.getPath() = this)
    or
    exists(FileReaderWriters fw | fw.getToPath() = this)
    or
    exists(FileWriters fw | fw.getPath() = this)
    or
    exists(DirectoryWriters dw | dw.getPath() = this)
    or
    exists(DirectoryReaders dr | dr.getPath() = this)
  }
}
