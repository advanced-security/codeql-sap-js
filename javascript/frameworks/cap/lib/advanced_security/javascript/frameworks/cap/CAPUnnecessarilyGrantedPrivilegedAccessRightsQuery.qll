import javascript
import advanced_security.javascript.frameworks.cap.CDS

/**
 * A reference to an entity that is defined in this application.
 */
class LocalEntityReference instanceof EntityReference {
  LocalEntityReference() { not this instanceof RemoteEntityReference }

  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }

  predicate hasRestrictedAccessControl() {
    exists(RestrictCondition restrict |
      restrict =
        this.(EntityReference).getCqlDefinition().getRestrictAnnotation().getARestrictCondition()
    |
      not restrict.grantsToAnyone(_)
    )
  }
}

/**
 * A reference to an entity that is not defined in this application and
 * read from a service instance that is looked up with the name defined in
 * package.json.
 */
class RemoteEntityReference instanceof EntityReference {
  RemoteEntityReference() { not exists(this.getCqlDefinition()) }

  string toString() { result = super.toString() }

  Location getLocation() { result = super.getLocation() }
}

abstract class PrivilegedUserInstance extends DataFlow::Node { }

class CdsUserPrivilegedProperty extends PrivilegedUserInstance instanceof PropRead {
  CdsUserPrivilegedProperty() {
    exists(CdsUser cdsUser, PropRead cdsUserRef |
      cdsUserRef = cdsUser.getInducingNode() and
      cdsUserRef.flowsTo(this.getBase()) and
      this.getPropertyName() = "Privileged"
    )
  }
}

class CustomPrivilegedUser extends ClassNode {
  CustomPrivilegedUser() {
    exists(CdsUser cdsUser | this.getASuperClassNode() = cdsUser.asSource()) and
    exists(FunctionNode init |
      init = this.getInstanceMethod("is") and
      forall(Expr expr | expr = init.asExpr().(Function).getAReturnedExpr() |
        expr.mayHaveBooleanValue(true)
      )
    )
  }
}

class CustomPrivilegedUserInstance extends PrivilegedUserInstance, NewNode {
  CustomPrivilegedUserInstance() {
    exists(CustomPrivilegedUser customPrivilegedUserClass |
      this = customPrivilegedUserClass.getAnInstantiation()
    )
  }
}
