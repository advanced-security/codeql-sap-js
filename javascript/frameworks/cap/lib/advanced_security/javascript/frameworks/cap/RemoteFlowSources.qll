import javascript
import advanced_security.javascript.frameworks.cap.CDS

/**
 * The request parameter of a handler belonging to a service that is exposed to
 * a protocol. e.g. All parameters named `req` is captured in the below example.
 * ``` javascript
 * // srv/service1.js
 * module.exports = class Service1 extends cds.ApplicationService {
 *   this.on("SomeEvent", "SomeEntity", (req) => { ... });
 *   this.before("SomeEvent", "SomeEntity", (req, next) => { ... });
 * }
 * ```
 * ``` cds
 * // srv/service1.cds
 * service Service1 @(path: '/service-1') { ... }
 * ```
 *
 * NOTE: CDS extraction can fail for various reasons, and if so the detection
 * logic falls back on overapproximating on the parameters and assume they are
 * exposed.
 */
class HandlerParameterOfExposedService extends HandlerParameter {
  HandlerParameterOfExposedService() {
    /* 1. The CDS definition is there and we can determine it is exposed. */
    this.getHandler().getHandlerRegistration().getService().getDefinition().isExposed()
    or
    /*
     * 2. (Fallback) The CDS definition is not there, so no precise service definition
     * is known.
     */

    not exists(
      this.getHandler().getHandlerRegistration().getService().getDefinition().getCdsDeclaration()
    )
  }
}

/**
 * Reads of property belonging to a request parameter that is exposed to a protocol.
 * It currently models the following access paths:
 * - `req.data` (from `cds.Event.data`)
 * - `req.params` (from `cds.Request.params`)
 * - `req.headers` (from `cds.Event.headers`)
 * - `req.http.req.*` (from `cds.EventContext.http.req`)
 * - `req.id` (from `cds.EventContext.id`)
 *
 * Note that `req.http.req` has type `require("@express").Request`, so their uses are
 * completely identical. Subsequently, the models for this access path follow Express'
 * API descriptions (as far back as 3.x). Also see `Express::RequestInputAccess`,
 * `Express::RequestHeaderAccess`, and `Express::RequestBodyAccess` of the standard
 * library.
 */
class UserProvidedPropertyReadOfHandlerParameterOfExposedService extends RemoteFlowSource {
  HandlerParameterOfExposedService handlerParameterOfExposedService;

  UserProvidedPropertyReadOfHandlerParameterOfExposedService() {
    /* 1. `req.(data|params|headers|id)` */
    this = handlerParameterOfExposedService.getAPropertyRead(["data", "params", "headers", "id"])
    or
    /* 2. APIs stemming from `req.http.req`: Defined by Express.js */
    exists(PropRead reqHttpReq |
      reqHttpReq = handlerParameterOfExposedService.getAPropertyRead("http").getAPropertyRead("req")
    |
      this = reqHttpReq.getAPropertyRead(["originalUrl", "hostname"]) or
      this =
        reqHttpReq
            .getAPropertyRead(["query", "body", "params", "headers", "cookies"])
            .getAPropertyRead() or
      this = reqHttpReq.getAMemberCall(["get", "is", "header", "param"])
    )
  }

  HandlerParameterOfExposedService getHandlerParameter() {
    result = handlerParameterOfExposedService
  }

  Handler getHandler() { result = handlerParameterOfExposedService.getHandler() }

  override string getSourceType() {
    result =
      "Tainted property read of the request parameter of an event handler belonging to an exposed service"
  }
}
