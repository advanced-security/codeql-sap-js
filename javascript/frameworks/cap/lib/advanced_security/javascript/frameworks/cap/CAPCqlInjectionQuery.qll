import javascript
import semmle.javascript.security.dataflow.SqlInjectionCustomizations
import advanced_security.javascript.frameworks.cap.CQL
import advanced_security.javascript.frameworks.cap.RemoteFlowSources
import advanced_security.javascript.frameworks.cap.dataflow.FlowSteps

class CqlClauseWithStringConcatParameter instanceof CqlClause {
  CqlClauseWithStringConcatParameter() {
    exists(DataFlow::Node queryParameter |
      (
        if this instanceof CqlInsertClause or this instanceof CqlUpsertClause
        then
          queryParameter = this.getArgument().flow() or
          queryParameter = this.getArgument().flow().(SourceNode).getAPropertyWrite().getRhs()
        else queryParameter = this.getArgument().flow()
      ) and
      exists(StringConcatenation::getAnOperand(queryParameter))
    )
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlShortcutMethodCallWithStringConcat instanceof CqlShortcutMethodCall {
  CqlShortcutMethodCallWithStringConcat() {
    exists(StringConcatenation::getAnOperand(super.getAQueryParameter()))
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlClauseParserCallWithStringConcat instanceof CqlClauseParserCall {
  CqlClauseParserCallWithStringConcat() {
    exists(StringConcatenation::getAnOperand(super.getCdlString()))
  }

  Location getLocation() { result = super.getLocation() }

  string toString() { result = super.toString() }
}

class CqlInjectionConfiguration extends TaintTracking::Configuration {
  CqlInjectionConfiguration() { this = "CQL injection from untrusted data" }

  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }

  override predicate isSink(DataFlow::Node node) {
    exists(CqlRunMethodCall cqlRunMethodCall |
      node = cqlRunMethodCall.(CqlRunMethodCall).getAQueryParameter()
    )
    or
    exists(CqlShortcutMethodCallWithStringConcat queryRunnerCall |
      node = queryRunnerCall.(CqlQueryRunnerCall).getAQueryParameter()
    )
    or
    exists(AwaitExpr await, CqlClauseWithStringConcatParameter cqlClauseWithStringConcat |
      node = await.flow() and
      await.getOperand() = cqlClauseWithStringConcat.(CqlClause).asExpr()
    )
  }

  override predicate isSanitizer(DataFlow::Node node) { node instanceof SqlInjection::Sanitizer }

  override predicate isAdditionalTaintStep(DataFlow::Node start, DataFlow::Node end) {
    exists(CqlClauseParserCallWithStringConcat cqlParseCallWithStringConcat |
      start = cqlParseCallWithStringConcat.(CqlClauseParserCall).getAnArgument() and
      end = cqlParseCallWithStringConcat
    )
    or
    exists(CqlClause cqlClause |
      start = cqlClause.getArgument().flow() and
      end = cqlClause.flow()
    )
  }
}
