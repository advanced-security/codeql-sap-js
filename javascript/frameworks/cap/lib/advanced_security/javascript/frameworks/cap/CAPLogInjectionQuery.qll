import javascript
import semmle.javascript.dataflow.DataFlow
import semmle.javascript.security.dataflow.LogInjectionQuery
import advanced_security.javascript.frameworks.cap.RemoteFlowSources
import advanced_security.javascript.frameworks.cap.CDS
import advanced_security.javascript.frameworks.cap.dataflow.DataFlow

/**
 * A logger obtained by a call to `log` on a CDS facade. Each logger is associated with
 * its unique name.
 */
class CdsLogger extends MethodCallNode {
  string name;

  CdsLogger() {
    exists(CdsFacade cds |
      this = cds.getMember("log").getACall() and
      name = this.getArgument(0).getALocalSource().asExpr().(StringLiteral).getValue()
    )
  }

  string getName() { result = name }
}

class ConstantOnlyTemplateLiteral extends TemplateLiteral {
  ConstantOnlyTemplateLiteral() {
    forall(Expr e | e = this.getAnElement() | e instanceof TemplateElement)
  }
}

/**
 * Arguments of calls to `cds.log.{trace, debug, info, log, warn, error}`.
 */
class CdsLogSink extends DataFlow::Node {
  CdsLogSink() {
    exists(CdsLogger log, MethodCallNode loggingMethod |
      this = loggingMethod.getAnArgument() and
      loggingMethod.getMethodName() = ["trace", "debug", "info", "log", "warn", "error"] and
      not this.asExpr() instanceof Literal and
      not this.asExpr() instanceof ConstantOnlyTemplateLiteral and
      loggingMethod.getReceiver().getALocalSource() = log
    )
  }
}

class CAPLogInjectionConfiguration extends LogInjectionConfiguration {
  CAPLogInjectionConfiguration() { this = "CAP Log Injection" }

  override predicate isSource(DataFlow::Node start) {
    super.isSource(start)
    or
    start instanceof RemoteFlowSource
  }

  override predicate isBarrier(DataFlow::Node node) {
    exists(HandlerParameterData handlerParameterData |
      node = handlerParameterData and
      not handlerParameterData.getType() = ["cds.String", "cds.LargeString"]
    )
  }

  override predicate isSink(DataFlow::Node end) { end instanceof CdsLogSink }
}
