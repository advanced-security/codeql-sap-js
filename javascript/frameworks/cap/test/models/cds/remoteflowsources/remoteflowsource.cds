namespace sap.capire.test;

entity Test {

   key id:Integer;
}