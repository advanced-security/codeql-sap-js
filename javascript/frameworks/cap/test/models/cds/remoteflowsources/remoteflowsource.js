const loki = require('lokijs')
const db = new loki('DB')
const testDB = db.addCollection('Test')

module.exports = srv => {
    srv.before('CREATE', 'Test', req => { //source
      const obj = testDB.insert({ test: '' })
      req.data.id = obj.$loki
    })
}