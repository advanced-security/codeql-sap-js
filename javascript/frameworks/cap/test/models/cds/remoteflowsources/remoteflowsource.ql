import javascript
import advanced_security.javascript.frameworks.cap.RemoteFlowSources

from RemoteFlowSource source
select source
