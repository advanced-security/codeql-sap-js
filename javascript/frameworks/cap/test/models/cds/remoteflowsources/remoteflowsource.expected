| remoteflowsource.js:6:34:6:36 | req |
